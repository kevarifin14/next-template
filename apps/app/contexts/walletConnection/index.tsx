import { WalletAdapterNetwork } from "@solana/wallet-adapter-base";
import {
  ConnectionProvider,
  WalletProvider,
} from "@solana/wallet-adapter-react";
import {
  PhantomWalletAdapter,
  SlopeWalletAdapter,
  SolflareWalletAdapter,
  SolletExtensionWalletAdapter,
  SolletWalletAdapter,
} from "@solana/wallet-adapter-wallets";
import { clusterApiUrl } from "@solana/web3.js";
import { useMemo } from "react";

export default function WalletConnectionProvider({ children }) {
  const network =
    process.env.NODE_ENV === "development"
      ? WalletAdapterNetwork.Devnet
      : WalletAdapterNetwork.Mainnet;

  const endpoint = useMemo(() => clusterApiUrl(network), [network]);

  const wallets = useMemo(
    () => [
      new PhantomWalletAdapter(),
      new SlopeWalletAdapter(),
      new SolflareWalletAdapter(),
      new SolletWalletAdapter({ network }),
      new SolletExtensionWalletAdapter({ network }),
    ],
    [network]
  );

  return (
    <ConnectionProvider endpoint={endpoint}>
      <WalletProvider wallets={wallets} autoConnect>
        {children}
      </WalletProvider>
    </ConnectionProvider>
  );
}
