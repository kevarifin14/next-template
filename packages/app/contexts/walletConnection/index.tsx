import { ConnectionProvider, WalletProvider } from '@solana/wallet-adapter-react';
import { getPhantomWallet, getSolletWallet } from '@solana/wallet-adapter-wallets';
import { useMemo } from 'react';

export default function WalletConnectionProvider({ children }) {
  const wallets = useMemo(() => [
    getPhantomWallet(),
    getSolletWallet(),
  ], []);

  return (
    <ConnectionProvider endpoint="http://localhost:8899">
      <WalletProvider wallets={wallets} autoConnect>
        {children}
      </WalletProvider>
    </ConnectionProvider>
  );
}
