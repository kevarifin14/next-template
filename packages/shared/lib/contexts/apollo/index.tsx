/* istanbul ignore file */
import {
  ApolloClient,
  HttpLink,
  InMemoryCache,
  ApolloProvider,
  ApolloLink,
} from "@apollo/client";

export const createHasuraClient = () => {
  const httpLink = new HttpLink({
    uri: `${process.env.NEXT_PUBLIC_APP_URL}/api/graphql`,
  });

  const link = ApolloLink.from([httpLink]);

  const cache = new InMemoryCache();
  return new ApolloClient({ link, cache });
};

export const client = createHasuraClient();

export function ApolloClientProvider({ children }) {
  return <ApolloProvider client={client}>{children}</ApolloProvider>;
}
