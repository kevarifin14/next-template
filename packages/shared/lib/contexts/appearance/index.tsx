import React, {
  createContext, Dispatch, ReactNode, SetStateAction, useContext, useState,
} from 'react';

type AppearanceContextProps = {
  dark: boolean,
  setDark: Dispatch<SetStateAction<boolean>>,
};

export const AppearanceContext = createContext<AppearanceContextProps>(null);

export const useAppearanceContext = () => useContext(AppearanceContext);

type AppearanceProviderProps = {
  dark?: boolean,
  children: ReactNode,
};

export function AppearanceProvider({
  dark: defaultDark, children,
}: AppearanceProviderProps) {
  const [dark, setDark] = useState(defaultDark);

  return (
    <AppearanceContext.Provider value={{ dark, setDark }}>
      <div className={dark ? 'dark' : ''}>
        {children}
      </div>
    </AppearanceContext.Provider>
  );
}
