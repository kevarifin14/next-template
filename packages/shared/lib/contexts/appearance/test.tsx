import { render, fireEvent } from '@testing-library/react';
import React from 'react';

import { AppearanceProvider, useAppearanceContext } from '.';

describe('Apprearance Context', () => {
  const AppearanceConsumer = () => {
    const { dark, setDark } = useAppearanceContext();
    return (
      <>
        <button onClick={() => setDark((prev) => !prev)}>Toggle Dark Mode</button>
        {dark ? 'dark' : 'light'}
      </>
    );
  };

  const wrapper = ({ children }) => <AppearanceProvider dark>{children}</AppearanceProvider>;

  it('matches snapshot', () => {
    const { baseElement } = render(<AppearanceConsumer />, { wrapper });
    expect(baseElement).toMatchSnapshot();
  });

  it('toggles dark mode', () => {
    const { getByRole, queryByText } = render(<AppearanceConsumer />, { wrapper });

    const toggleDarkModeButton = getByRole('button');

    expect(queryByText('dark')).toBeTruthy();

    fireEvent.click(toggleDarkModeButton);

    expect(queryByText('light')).toBeTruthy();
  });
});
