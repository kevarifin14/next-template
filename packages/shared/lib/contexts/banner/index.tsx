/* istanbul ignore file */
import { Button } from "../../components/Button";
import { event } from "../../utils/ga";
import React, {
  createContext,
  Dispatch,
  ReactNode,
  SetStateAction,
  useContext,
  useState,
} from "react";
import { HiSpeakerphone, HiX } from "react-icons/hi";

type IBanner = {
  message: string;
  cta: string;
  key: string;
  ctaOnClick: () => void;
  closable?: boolean;
};

type BannerContextProps = {
  banner: IBanner;
  setBanner: Dispatch<SetStateAction<IBanner>>;
};

export const BannerContext = createContext<BannerContextProps>(null);

export const useBannerContext = () => useContext(BannerContext);

type BannerProviderProps = {
  children: ReactNode;
};

export function BannerProvider({ children }: BannerProviderProps) {
  const [banner, setBanner] = useState<IBanner>(null);

  const handleClose = () => setBanner(null);

  const handleOnClick = () => {
    banner.ctaOnClick();
    event("call to action", `cta - banner`, banner.key);
  };

  return (
    <BannerContext.Provider value={{ banner, setBanner }}>
      {children}

      {banner && (
        <div className="sticky bottom-0 inset-x-0 pb-2 px-2 sm:pb-5">
          <div className="flex justify-between max-w-5xl mx-auto bg-primary p-3 sm:px-6 rounded-md">
            <div className="flex items-center">
              <HiSpeakerphone
                className="h-6 w-6 text-white"
                aria-hidden="true"
              />
              <p className="ml-3 font-medium sm:text-lg text-white">
                {banner.message}
              </p>
            </div>

            <div className="flex justify-end items-center space-x-1 sm:space-x-2">
              <Button
                type="white"
                onClick={handleOnClick}
                className="hidden sm:flex"
              >
                {banner.cta}
              </Button>

              <Button
                type="white"
                size="sm"
                onClick={handleOnClick}
                className="sm:hidden"
              >
                {banner.cta}
              </Button>

              {banner.closable && (
                <button onClick={handleClose}>
                  <HiX className="text-white text-xl font-black" />
                </button>
              )}
            </div>
          </div>
        </div>
      )}
    </BannerContext.Provider>
  );
}
