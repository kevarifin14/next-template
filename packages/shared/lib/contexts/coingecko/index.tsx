import React, { useContext, useEffect, useState, createContext } from "react";

export const COINGECKO_POOL_INTERVAL = 1000 * 60; // 60 sec
export const COINGECKO_API = "https://api.coingecko.com/api/v3/simple/price";

export interface CoingeckoContextState {
  solPrice: number;
}

export const solToUSD = async (): Promise<number> => {
  const url = `${COINGECKO_API}?ids=solana&vs_currencies=usd`;
  const resp = await window.fetch(url).then((r) => r.json());
  return resp.solana.usd;
};

const CoingeckoContext = createContext<CoingeckoContextState>(null);

export function CoingeckoProvider({ children = null as any }) {
  const [solPrice, setSolPrice] = useState<number>(0);

  useEffect(() => {
    let timerId = 0;

    const startTimer = () => {
      timerId = window.setTimeout(async () => {
        queryPrice();
      }, COINGECKO_POOL_INTERVAL);
    };

    const queryPrice = async () => {
      const solprice = await solToUSD();
      setSolPrice(solprice);
      startTimer();
    };

    queryPrice();

    return () => {
      clearTimeout(timerId);
    };
  }, [setSolPrice]);

  return (
    <CoingeckoContext.Provider value={{ solPrice }}>
      {children}
    </CoingeckoContext.Provider>
  );
}

export const useCoingecko = () => {
  const context = useContext<CoingeckoContextState>(CoingeckoContext);
  return context;
};

export const useSolPrice = () => {
  const { solPrice } = useCoingecko();

  return solPrice;
};
