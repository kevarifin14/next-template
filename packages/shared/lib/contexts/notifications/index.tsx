import { Transition } from '@headlessui/react';
import React, {
  createContext, useState, useCallback, useContext, useEffect,
} from 'react';
import { IconType } from 'react-icons';
import { HiX } from 'react-icons/hi';
import { v4 as uuid } from 'uuid';

import { CheckCircleIcon, ExclamationCircleIcon } from '../../components/Icon';

export interface INotification {
  id: string,
  title: string,
  description?: string,
  icon?: IconType,
}

export type NotificationContextProps = {
  notifications: INotification[],
  addNotification: (notification: Omit<INotification, 'id'>) => void,
  addSuccessNotification: (notification: Omit<INotification, 'id' | 'icon'>) => void,
  addErrorNotification: (notification: Omit<INotification, 'id' | 'icon'>) => void,
  removeNotification: (notification: INotification) => void,
};

export const NotificationContext = createContext<NotificationContextProps>(null);

export function useNotifications() {
  const notificationContext = useContext(NotificationContext);
  return notificationContext;
}

type NotificationProps = {
  notification: INotification,
};

export function Notification({ notification }: NotificationProps) {
  const { removeNotification } = useNotifications();
  const { id, title, description } = notification;

  useEffect(() => {
    const interval = setInterval(() => removeNotification(notification), 4000);
    return () => clearInterval(interval);
  }, [removeNotification, id]);

  const handleRemoveNotification = () => {
    removeNotification(notification);
  };

  return (
    <Transition
      key={id}
      show
      enter="transform ease-out duration-300 transition"
      enterFrom="translate-y-2 opacity-0 sm:translate-y-0 sm:translate-x-2"
      enterTo="translate-y-0 opacity-100 sm:translate-x-0"
      leave="transition ease-in duration-100"
      leaveFrom="opacity-100"
      leaveTo="opacity-0"
      className="max-w-sm w-full bg-white dark:bg-dark-light shadow-lg rounded-lg pointer-events-auto overflow-hidden z-50"
    >
      <div className="p-4 prose dark:prose-light">
        <div className="flex items-start">
          <div className="flex-shrink-0">
            {notification.icon && <notification.icon />}
          </div>
          <div className="ml-3 w-0 flex-1 pt-0.5">
            <p className="text-sm font-medium text-gray-900 dark:text-white m-0">
              {title}
            </p>
            <p className="mt-1 text-sm text-gray-500 m-0">
              {description}
            </p>
          </div>
          <div className="ml-4 flex-shrink-0 flex">
            <button
              className="rounded-md inline-flex text-gray-400 hover:text-gray-500 focus:outline-none"
              onClick={handleRemoveNotification}
            >
              <span className="sr-only">Close</span>
              <HiX className="h-5 w-5" />
            </button>
          </div>
        </div>
      </div>
    </Transition>
  );
}
export function NotificationsProvider({ children }) {
  const [notifications, setNotifications] = useState<INotification[]>([]);

  const addNotification = useCallback((notification: INotification) => {
    setNotifications((prevNotifications) => [
      ...prevNotifications,
      { ...notification, id: uuid() },
    ]);
  }, [setNotifications]);

  const addSuccessNotification = useCallback((notification: INotification) => {
    setNotifications((prevNotifications) => [
      ...prevNotifications,
      {
        ...notification,
        icon: CheckCircleIcon,
        id: uuid(),
      },
    ]);
  }, [setNotifications]);

  const addErrorNotification = useCallback((notification: INotification) => {
    setNotifications((prevNotifications) => [
      ...prevNotifications,
      {
        ...notification,
        icon: ExclamationCircleIcon,
        id: uuid(),
      },
    ]);
  }, [setNotifications]);

  const removeNotification = useCallback((notification) => {
    setNotifications(
      (prevNotifications) => prevNotifications.filter(({ id }) => id !== notification.id),
    );
  }, [setNotifications]);

  return (
    <NotificationContext.Provider
      value={{
        notifications,
        addNotification,
        addSuccessNotification,
        addErrorNotification,
        removeNotification,
      }}
    >
      <div className="fixed max-w-sm w-full right-0 top-0 flex flex-col justify-end items-start space-y-4 pointer-events-none z-30 p-4">
        {notifications.map((notification) => (
          <Notification key={notification.id} notification={notification} />))}
      </div>
      {children}
    </NotificationContext.Provider>
  );
}
