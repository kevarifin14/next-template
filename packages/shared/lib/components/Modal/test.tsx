import { ConfirmModal, Modal } from ".";
import { AppearanceProvider } from "../../contexts/appearance";
import { render, fireEvent, act } from "@testing-library/react";
import React from "react";

describe("Modal", () => {
  const mockChildren = "mockChildren";

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("matches snapshot", () => {
    const { baseElement } = render(<Modal open>{mockChildren}</Modal>);
    expect(baseElement).toMatchSnapshot();
  });

  it("renders Modal with content", () => {
    const { getByText } = render(<Modal open>{mockChildren}</Modal>);
    expect(getByText(mockChildren)).toContainHTML(mockChildren);
  });

  it("renders does not render content when closed", () => {
    const { queryByText } = render(<Modal open={false}>{mockChildren}</Modal>);
    expect(queryByText(mockChildren)).toBeNull();
  });

  it("fires onClose function", () => {
    const mockOnClose = jest.fn();
    const { getByRole } = render(
      <Modal open onClose={mockOnClose}>
        {mockChildren}
      </Modal>
    );

    const closeXIcon = getByRole("button");
    fireEvent.click(closeXIcon);

    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  it("works without an onClose function", () => {
    const { getByRole } = render(<Modal open>{mockChildren}</Modal>);

    const closeXIcon = getByRole("button");
    fireEvent.click(closeXIcon);
  });

  it("hides close button when not closable", () => {
    const { queryByRole } = render(
      <Modal open closable={false}>
        {mockChildren}
      </Modal>
    );
    expect(queryByRole("button")).toBeNull();
  });

  it("adds dark class", () => {
    const { container } = render(
      <AppearanceProvider dark>
        <Modal open>{mockChildren}</Modal>
      </AppearanceProvider>
    );
    expect(container.firstChild).toHaveClass("dark");
  });
});

describe("ConfirmModal", () => {
  it("matches snapshot", () => {
    const { baseElement } = render(<ConfirmModal title="" onOk={() => {}} />);
    expect(baseElement).toMatchSnapshot();
  });

  it("fires onOk function", async () => {
    const mockOnOk = jest.fn();
    const { getByRole } = render(<ConfirmModal title="" onOk={mockOnOk} />);

    const okButton = getByRole("button", { name: "Ok" });

    await act(async () => {
      fireEvent.click(okButton);
    });

    expect(mockOnOk).toHaveBeenCalledTimes(1);
  });

  it("onCancel closes modal", async () => {
    const mockOnOk = jest.fn();
    const { getByRole } = render(
      <ConfirmModal title="confirm modal" onOk={mockOnOk} />
    );

    const cancelButton = getByRole("button", { name: "Cancel" });

    await act(async () => {
      fireEvent.click(cancelButton);
    });
  });
});
