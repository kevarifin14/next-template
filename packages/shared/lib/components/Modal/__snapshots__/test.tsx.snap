// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal matches snapshot 1`] = `
<body>
  <div />
  <div
    id="headlessui-portal-root"
  >
    <div>
      <div
        aria-modal="true"
        class="fixed z-10 inset-0 overflow-y-auto"
        id="headlessui-dialog-1"
        role="dialog"
      >
        <div
          class="flex items-end justify-center min-h-screen pt-4 px-4 pb-20 text-center sm:block sm:p-0"
        >
          <div
            aria-hidden="true"
            class="fixed inset-0 bg-gray-500 bg-opacity-75 transition-opacity"
            id="headlessui-dialog-overlay-4"
          />
          <span
            aria-hidden="true"
            class="hidden sm:inline-block sm:align-middle sm:h-screen"
          >
            â€‹
          </span>
          <div
            class="inline-block align-bottom bg-white rounded-lg p-4 text-left overflow-hidden shadow-xl transform transition-all sm:my-8 sm:align-middle sm:max-w-sm sm:w-full"
          >
            <div
              class="flex justify-between"
            >
              <h1 />
              <button
                class="bg-white rounded-md text-gray-400 hover:text-gray-500 focus:outline-none"
                tabindex="0"
              >
                <svg
                  aria-hidden="true"
                  class="h-6 w-6"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 20 20"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </div>
            <div
              class="py-4"
            >
              mockChildren
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
